qualtran/_infra/controlled.py

```diff
@@ -11,59 +11,292 @@
+
+        cv = self.cvs[soq.idx]
+        return TextBox(f'{cv}')
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, CtrlSpec):
+            return False
+
+        return (
+            other._bitsize == self._bitsize
+            and other.shape == self.shape
+            and other._cvs_tuple == self._cvs_tuple
+        )
+
+    def __hash__(self):
+        if self._hash is None:
+            self._hash = hash((self.bitsize, self.shape, self._cvs_tuple))
+        return self._hash
+
+    def __repr__(self) -> str:
+        return f'CtrlSpec({self.bitsize!r}, {self.cvs!r})'
 
     def __str__(self) -> str:
-        return f'C[{self.subbloq}]'
+        return self.__repr__()
+
+
+class AddControlledT(Protocol):
+    def __call__(
+        self, bb: 'BloqBuilder', ctrl_soqs: Sequence['SoquetT'], in_soqs: Dict[str, 'SoquetT']
+    ) -> Tuple[Iterable['SoquetT'], Iterable['SoquetT']]:
+        """The signature for the `add_controlled` callback part of `ctrl_system`.
+
+        Crucially, `ctrl_soqs` must be compatible with the ControlSpec, specifically with
+        the control registers implied by `activation_function_dtypes`.
+        """
+
+
+@frozen
+class Controlled(Bloq):
+    """A controlled version of `subbloq`.
+
+    This meta-bloq is part of the 'controlled' protocol. As a default fallback,
+    we wrap any bloq without a custom controlled version in this meta-bloq.
+
+    Users should likely not use this class directly. Prefer using `bloq.controlled(ctrl_spec)`,
+    which may return a tailored Bloq that is controlled in the desired way.
+
+    Args:
+        subbloq: The bloq we are controlling.
+        ctrl_spec: The specification for how to control the bloq.
+    """
+
+    subbloq: Bloq
+    ctrl_spec: AbstractCtrlSpec
+
+    @classmethod
+    def make_ctrl_system(
+        cls, bloq: Bloq, ctrl_spec: AbstractCtrlSpec
+    ) -> Tuple[Bloq, AddControlledT]:
+        """A factory method for creating both the Controlled and the adder function.
+
+        See `Bloq.get_ctrl_system`.
+        """
+        cb = cls(subbloq=bloq, ctrl_spec=ctrl_spec)
+        ctrl_reg_names = cb.ctrl_reg_names
+
+        def add_controlled(
+            bb: 'BloqBuilder', ctrl_soqs: Sequence['SoquetT'], in_soqs: Dict[str, 'SoquetT']
+        ) -> Tuple[Iterable['SoquetT'], Iterable['SoquetT']]:
+            in_soqs |= dict(zip(ctrl_reg_names, ctrl_soqs))
+            new_out_d = bb.add_d(cb, **in_soqs)
+
+            # We need to pluck out the `ctrl_soq` from the new out_soqs
+            # dictionary, so we can correspond it to `old_out_soqs`.
+            ctrl_soqs = tuple(new_out_d.pop(creg_name) for creg_name in ctrl_reg_names)
+            return ctrl_soqs, new_out_d.values()
+
+        return cb, add_controlled
 
     @cached_property
-    def signature(self) -> Signature:
-        return Signature((Register(name="control", bitsize=1),) + tuple(self.subbloq.signature))
+    def ctrl_reg_names(self) -> Sequence[str]:
```

Comment:
Does it make sense to make this an independent function outside of `Controlled` bloq?

Reply:
why?

Reply:
I'll factor it into a private helper function